internal/statemachine/classic_states.go

```diff
@@ -296,6 +311,101 @@ func (stateMachine *StateMachine) createChroot() error {
+	}
+
+	// now create the ppa sources.list files
+	tmpGPGDir, err := osMkdirTemp("/tmp", "ubuntu-image-gpg")
+	if err != nil {
+		return fmt.Errorf("Error creating temp dir for gpg imports: %s", err.Error())
+	}
+	defer osRemoveAll(tmpGPGDir)
+	for _, ppa := range classicStateMachine.ImageDef.Customization.ExtraPPAs {
+		ppaFileName, ppaFileContents := createPPAInfo(ppa,
+			classicStateMachine.ImageDef.Series)
+
+		ppaFile := filepath.Join(sourcesListD, ppaFileName)
+		ppaIO, err := osOpenFile(ppaFile, os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return fmt.Errorf("Error creating %s: %s", ppaFile, err.Error())
+		}
+		ppaIO.Write([]byte(ppaFileContents))
+		ppaIO.Close()
+
+		// Import keys either from the specified fingerprint or via the Launchpad API
+		keyFileName := strings.Replace(ppaFileName, ".sources", ".gpg", 1)
+		keyFilePath := filepath.Join(classicStateMachine.tempDirs.chroot,
+			"etc", "apt", "trusted.gpg.d", keyFileName)
+		err = importPPAKeys(ppa, tmpGPGDir, keyFilePath)
+		if err != nil {
+			return fmt.Errorf("Error retrieving signing key for ppa \"%s\": %s",
+				ppa.PPAName, err.Error())
+		}
+	}
+	if err := osRemoveAll(tmpGPGDir); err != nil {
+		return fmt.Errorf("Error removing temporary gpg directory \"%s\": %s", tmpGPGDir, err.Error())
+	}
+
+	return nil
+}
+
+// Install packages in the chroot environment. This is accomplished by
+// running commands to do the following:
+// 1. Mount /proc /sys and /dev in the chroot
+// 2. Run `apt update` in the chroot
+// 3. Run `apt install <package list>` in the chroot
+// 4. Unmount /proc /sys and /dev
+func (stateMachine *StateMachine) installPackages() error {
+	var classicStateMachine *ClassicStateMachine
+	classicStateMachine = stateMachine.parent.(*ClassicStateMachine)
+
+	// if any extra packages are specified, install them alongside the seeded packages
+	if classicStateMachine.ImageDef.Customization != nil {
+		for _, packageInfo := range classicStateMachine.ImageDef.Customization.ExtraPackages {
+			classicStateMachine.Packages = append(classicStateMachine.Packages,
+				packageInfo.PackageName)
+		}
+	}
+
+	// Slice used to store all the commands that need to be run
+	// to install the packages
+	var installPackagesCmds []*exec.Cmd
+
+	// mount some necessary partitions from the host in the chroot
+	mounts := []string{"/dev", "/proc", "/sys"}
+	var umounts []*exec.Cmd
+	for _, mount := range mounts {
+		mountCmd, umountCmd := mountFromHost(stateMachine.tempDirs.chroot, mount)
+		defer umountCmd.Run()
+		installPackagesCmds = append(installPackagesCmds, mountCmd)
+		umounts = append(umounts, umountCmd)
+	}
+
+	// generate the apt update/install commands and append them to the slice of commands
+	aptCmds := generateAptCmds(stateMachine.tempDirs.chroot, classicStateMachine.Packages)
+	installPackagesCmds = append(installPackagesCmds, aptCmds...)
+	installPackagesCmds = append(installPackagesCmds, umounts...) // don't forget to unmount!
```

Comment:
Same here. Might be the normal convention, but this means that we will run additional `umount` commands that will fail and not be needed? What is the usual thing to do in such cases?